app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt

```diff
@@ -450,17 +453,46 @@ class PatientSummaryScreenController @Inject constructor(
   }
 
   private fun showUpdatePhoneDialogIfRequired(events: Observable<UiEvent>): Observable<UiChange> {
-    val patientUuidStream = events
-        .ofType<PatientSummaryScreenCreated>()
+    val screenCreations = events.ofType<PatientSummaryScreenCreated>()
+
+    val showForInvalidPhone = screenCreations
         .map { it.patientUuid }
+        .switchMap { patientUuid ->
+          hasInvalidPhone(patientUuid)
+              .take(1)
+              .filter { invalid -> invalid }
+              .map { { ui: Ui -> ui.showUpdatePhoneDialog(patientUuid) } }
+        }
 
-    return Observables
-        .zip(patientUuidStream, configProvider.toObservable())
-        .filter { (_, config) -> config.isUpdatePhoneDialogEnabled }
-        .switchMap { (patientUuid, _) -> Observables.zip(phoneNumber(patientUuid), lastCancelledAppointment(patientUuid)) }
-        .filter { (number, appointment) -> appointment.updatedAt > number.updatedAt }
-        .take(1)
-        .map { (number) -> { ui: Ui -> ui.showUpdatePhoneDialog(number.patientUuid) } }
+    val showForMissingPhone = screenCreations
+        .filter { it.caller == SEARCH }
+        .map { it.patientUuid }
+        .switchMap { patientUuid ->
+          isMissingPhoneAndShouldBeReminded(patientUuid)
+              .take(1)
+              .filter { missing -> missing }
+              .flatMap {
+                missingPhoneReminderRepository
+                    .markReminderAsShownFor(patientUuid)
+                    .andThen(Observable.just({ ui: Ui -> ui.showAddPhoneDialog(patientUuid) }))
+              }
+        }
+
+    return showForInvalidPhone.mergeWith(showForMissingPhone)
+  }
+
+  private fun hasInvalidPhone(patientUuid: UUID): Observable<Boolean> {
+    return patientRepository.phoneNumber(patientUuid)
+        .filterAndUnwrapJust()
+        .zipWith(lastCancelledAppointment(patientUuid))
+        .map { (number, appointment) -> appointment.updatedAt > number.updatedAt }
+  }
+
+  private fun isMissingPhoneAndShouldBeReminded(patientUuid: UUID): Observable<Boolean> {
```

Comment:
@vinaysshenoy I'm going back to my older way of calculating when is a patient visiting for the 2nd time:

`When the patient summary screen is opened from search` + `the user has never been reminded before for this patient`.

I don't think it's possible to deduce this otherwise with just the existing data. We will have to persist another value.